core/libp2p/network/negotiator.hpp

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_NEGOTIATOR_HPP
+#define KAGOME_NEGOTIATOR_HPP
+
+#include <memory>
+
+#include "libp2p/basic/readwritecloser.hpp"
+#include "libp2p/peer/protocol.hpp"
+#include "libp2p/connection/stream.hpp"
+
+namespace libp2p::network {
+
+  /**
+   * @brief Negotiator is a component capable of reaching agreement over what
+   * protocols to use for inbound streams of communication.
+   * @see
+   * https://sourcegraph.com/github.com/libp2p/go-libp2p-core@consolidate-skeleton/-/blob/protocol/switch.go#L53
+   */
+  struct Negotiator {
+    virtual ~Negotiator() = default;
+
+    // TODO: come up with better name
+    struct Return {
+      peer::Protocol protocol;
+      std::function<connection::Stream::Handler> handler;
+    };
+
+    // NegotiateLazy will return the registered protocol handler to use
+    // for a given inbound stream, returning as soon as the protocol has been
+    // determined. Returns an error if negotiation fails.
+    //
+    // NegotiateLazy may return before all protocol negotiation responses have
+    // been written to the stream. This is in contrast to Negotiate, which will
+    // block until the Negotiator is finished with the stream.
+    virtual outcome::result<Return> negotiateLazy(
+        std::shared_ptr<basic::ReadWriteCloser> io) = 0;
```

Comment:
I think we should pass stream instead of ReadWriteCloser. There will be several classes that fulfill RWC interface, and not all of them should be compatible with negotiator, right?

Reply:
Stream is also RWC

Reply:
Yes, but we should not accept connection here